File: Data Structures/Trees/Code/Trees-1.cpp
#include<iostream>
using namespace std;

class Node{
    public:

    int data;
    Node* left;
    Node* right;

    Node(int data){
        data = data;
        left = NULL;
        right = NULL;
    }
};

Node* createTree(){
    int nodeData;
    cout<<"Enter Data: ";
    cin>>nodeData;

    if(nodeData == -1) return NULL;

    Node* root = new Node(nodeData);

    cout<<"Left Sub-tree for "<<nodeData<<endl;
    root->left = createTree();

    cout<<"Right Sub-tree for "<<nodeData<<endl;
    root->right = createTree();
    
    return root;
}

void inorder(){

}

void preorder(){

}

void postorder(){

}

int main(){
    Node* root = createTree();
    cout<<root;

    return 0;
}

File: Data Structures/Trees/Code/Trees-1.js
const input = require("readline-sync");

class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

let root = createTree();
console.log("Pre-order");
preorder(root);
console.log("In-order");
inorder(root);
console.log("Post-order");
postorder(root);

//  Returns the root Node of the new tree
function createTree(){
    let nodeData = input.question("Enter data: ");
    if(nodeData == -1) return null;

    let root = new Node(nodeData);

    console.log(`Left sub-tree for ${nodeData}: `);
    root.left = createTree();

    console.log(`Right sub-tree for ${nodeData}: `);
    root.right = createTree();

    return root;
}

function preorder(root){
    if(root == null) return;

    console.log(root.data+" ");
    preorder(root.left);
    preorder(root.right);
}

function postorder(root){
    if(root == null) return;

    postorder(root.left);
    postorder(root.right);    
    console.log(root.data+" ");
}

function inorder(root){
    if(root == null) return;

    inorder(root.left);
    console.log(root.data+" ");
    inorder(root.right);
}